src/Stack/Freeze.hs

```diff
@@ -0,0 +1,52 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Stack.Freeze
+    ( freeze
+    , FreezeOpts (..)
+    ) where
+
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import qualified Data.Yaml as Yaml
+import qualified RIO.ByteString as B
+import           Stack.Prelude
+import           Stack.Build.Target
+import           Stack.Types.Config
+import           Stack.Types.Package
+import           Stack.Build.Source
+
+
+data FreezeOpts = FreezeOpts
+    { freezeTargets :: [Text]
+    -- ^ stack TARGETs to trace dependencies for
+    , freezeFlags :: !(Map (Maybe PackageName) (Map FlagName Bool))
+    -- ^ Flags to apply when calculating dependencies
+    , freezeExtraTargets :: !ExtraTargets
+    -- ^ Like the "--test" / "--bench" flag for build, affects the meaning of 'dotTargets'. 
+    }
+
+freeze :: HasEnvConfig env => FreezeOpts -> RIO env ()
+freeze freezeOpts = do
```

Comment:
Instead of working at the fully-parsed level of a source map, it would be better to work on the original `stack.yaml` file and any custom snapshots it refers to. I can step through that with you on a call if that would be helpful.

Reply:
I just couldn't find that yet and checking out code of `stack dot` was easy

Reply:
After having a testcase changing implementation should be even easier